src/plugin/src/main/java/com/defenseunicorns/uds/keycloak/plugin/authentication/RequireGroupAuthenticator.java

```diff
@@ -0,0 +1,148 @@
+package com.defenseunicorns.uds.keycloak.plugin.authentication;
+
+import java.util.Optional;
+import org.jboss.logging.Logger;
+import org.keycloak.authentication.AuthenticationFlowContext;
+import org.keycloak.authentication.AuthenticationFlowError;
+import org.keycloak.authentication.Authenticator;
+import org.keycloak.models.ClientModel;
+import org.keycloak.models.GroupModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.UserModel;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+/**
+ * Simple {@link Authenticator} that checks if a user is a member of a given {@link GroupModel Group}.
+ */
+public class RequireGroupAuthenticator implements Authenticator {
+
+    private static final Logger LOGGER = Logger.getLogger(RequireGroupAuthenticator.class.getName());
+
+    @Override
+    public void authenticate(final AuthenticationFlowContext context) {
+        UserModel user = context.getUser();
+        if (user == null) {
+            logAndFail(context, "Invalid user", AuthenticationFlowError.INVALID_USER);
+            return;
+        }
+
+        ClientModel client = context.getAuthenticationSession().getClient();
+        String groupsAttribute = client.getAttribute("uds.core.groups");
+
+        if (groupsAttribute == null || groupsAttribute.isEmpty()) {
+            LOGGER.infof("No groups detected for client %s", client.getName());
+            context.success();
+            return;
+        }
+
+        JsonNode groupsNode = parseGroupsAttribute(groupsAttribute);
+        if (groupsNode == null) {
+            logAndFail(context, "Failed to parse groups JSON", AuthenticationFlowError.INVALID_CLIENT_SESSION);
+            return;
+        }
+
+        handleGroupAuthentication(context, user, groupsNode);
+    }
+
+    private JsonNode parseGroupsAttribute(String groupsAttribute) {
+        ObjectMapper objectMapper = new ObjectMapper();
+        try {
+            return objectMapper.readTree(groupsAttribute).path("anyOf");
+        } catch (Exception e) {
+            LOGGER.errorf("Failed to parse groups JSON: %s", e.getMessage());
+            return null;
+        }
+    }
+
+    private void handleGroupAuthentication(AuthenticationFlowContext context, UserModel user, JsonNode groupsNode) {
+        if (!groupsNode.isArray() || groupsNode.size() == 0) {
+            logAndFail(context, "Groups attribute does not contain a valid anyOf array", AuthenticationFlowError.INVALID_CLIENT_SESSION);
+            return;
+        }
```

Comment:
Move the `groupsNode.size() > 0` check inside `parseGroupsAttribute()`. (and you won't need to test `isArray()` if you parse into a class)

Reply:
moving the size check into `parseGroupsAttribute` does require passing in the context to that method for the context.failure. Something i was attempting to contain in the `authenticate` method. I'm indifferent if you feel strongly about that.

Reply:
There's already a log message and `return null` for a different error (invalid json). We can handle this similarly I think?

`return null` on error. check for null in `authenticate()` and do the `context.failure()` there

Reply:
updated